static/js/code-highlighter.js

```diff
@@ -8,6 +8,198 @@
  * 3) Highlight lines when page loads, if window.location.hash exists
  */
 
+/**
+ * Decides what the document title should be.
+ */
+var DocumentTitler = new (class DocumentTitler {
+  constructor() {
+    this.originalTitle = document.title;
+    this.currentTitle = this.originalTitle;
+
+    this.stickyTitle = null;
+    this.selectionTitle = null;
+  }
+
+  updateTitle() {
+    let bestTitle;
+
+    // If we have a better title than the original title, use it, but making
+    // sure to include the original filename because this is important for
+    // finding the tab again in the awesomebar via its filename.
+    if (this.selectionTitle) {
+      bestTitle = `${this.selectionTitle} (${this.originalTitle})`;
+    } else if (this.stickyTitle) {
+      bestTitle = `${this.stickyTitle} (${this.originalTitle})`;
+    } else {
+      bestTitle = this.originalTitle;
+    }
+
+    // Debounce setting the title out of a fear of slowing down scrolling but
+    // not wanting to to do the legwork to figure out if this would pose a
+    // problem.  I am assuming this shouldn't result in synchronous reflows.
+    if (bestTitle === this.currentTitle) {
+      return;
+    }
+    document.title = this.currentTitle = bestTitle;
+  }
+
+  /**
+   * Recognizing that namespaces can be a little verbose and use up precious
+   * space, flatten things that look like namespaces to be a single-character
+   * delimited by a single colon.
+   *
+   *
+   * Good example transforms:
+   * - `Foo` => `Foo`
+   * - `Foo::Bar` => `Foo::Bar`
+   * - `mozilla::Foo::Bar` => `m:Foo::Bar`
+   * - `mozilla::dom::quota::Foo::Bar` => `m:d:q:Foo::Bar`
+   *
+   * Sketchy example transforms:
+   * - `mozilla::Foo` => `m:Foo`
+   * - `mozilla::dom::Foo` => `m:d:Foo`
+   *
+   * The sketchy transforms are sketchy because we're assuming that lowercase is
+   * indicative of a namespace.  Note that, similarly to the comments in
+   * `_findBestPrettySymbolInSourceLineElem`, this heuristic should simply be
+   * mooted by having the symbol dictionary provide a pre-computed concise
+   * pretty name for a symbol.  That mechanism can leverage knowing what is and
+   * isn't a namespace/class and also having global knowledge of the names that
+   * are in the codebase so that additional tokens can be used in cases where
+   * ambiguity exists, etc.
+   */
+  _shortenNamespaces(pretty) {
+    const pieces = pretty.split("::");
+    // Nothing to do if we don't have anything to split.
+    if (pieces.length < 2) {
+      return pretty;
+    }
+
+    // We want to use the last two components if they are both initial-caps,
+    // but not for the "mozilla::Foo" case, or "mozilla::dom::Foo" case, where
+    // we still want to collapse the lower-cased namespace.  In that case, we
+    // collapse everything but the last piece.
```

Comment:
I think I'd drop the collapsing of the `mozilla::` in `mozilla::Foo` (or at least special-case that instead).

Otherwise you incorrectly collapse classes that start with a lower-case letter when you're inside a method, which is very common in older code (I was trying e.g., `nsPresContext` in your instance).

Reply:
@emilio Ah, yeah, I see I already wrote a (better) check like this before in the fancy branch that has a [regexp /^ns|moz/](https://github.com/asutherland/mozsearch/blob/188fd265611efbe01fbc1b373551f067b5a35929/ui/src/grokysis/frontend/kb/symbol_info.js#L10) to decide things aren't namespaces.  But maybe a better hacky approach is checking that the entirety of the piece is 1) lowercase and 2) doesn't include underscores.

Reply:
My specific motivation here was for "mozilla::dom::indexeddb::HelperFunction" where "indexeddb" wasn't getting collapsed and was relatively long.  I could also probably just try and gather up the finite set of namespaces we have in the codebase, but I'm a little worried about people filing a bunch of bugs like "add my namespace to your big hacky regexp" when the real fix involves the structured records landing and some follow-ons.

Reply:
Yeah, checking that it's all lowercase seems like a reasonable compromise. We do have some mixed-case namespaces like `Foo_Bindings` etc for the DOM, but it's less common to explore autogenerated code I think.

Reply:
Implemented, thanks!